localstack/aws/connect.py

```diff
@@ -0,0 +1,310 @@
+
+LOG = logging.getLogger(__name__)
+
+#
+# Data transfer object
+#
+
+LOCALSTACK_REQUEST_PARAM_HEADER = "x-localstack-data"
+"""Request header which contains the data transfer object."""
+
+
+class LocalstackRequestParameters(TypedDict):
+    """
+    LocalStack Data Transfer Object.
+
+    This is sent with every internal request and contains any additional information
+    LocalStack might need for the purpose of policy enforcement. It is serialised
+    into text and sent in the request header.
+
+    Attributes can be added as needed. The keys should roughly correspond to:
+    https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_condition-keys.html
+    """
+
+    source_arn: str
+    """ARN of resource which is triggering the call"""
+
+    source_service: str
+    """Service principal where the call originates, eg. `ec2`"""
+
+    target_arn: str
+    """ARN of the resource being targeted."""
+
+
+def dump_dto(data: LocalstackRequestParameters) -> str:
+    # TODO@viren: Improve minification using custom JSONEncoder that use shortened keys
+
+    # To produce a compact JSON representation of DTO, remove spaces from separators
+    return json.dumps(data, separators=(",", ":"))
+
+
+def load_dto(data: str) -> LocalstackRequestParameters:
+    return json.loads(data)
+
+
+#
+# Client
+#
+
+# Names of arguments that can be passed to Boto API operation functions.
+# These have appropriate entries on the data transfer object.
+ARG_SOURCE_SERVICE = "_SourceService"
+ARG_SOURCE_ARN = "_SourceArn"
+ARG_TARGET_ARN = "_TargetArn"
+
+
+class ConnectFactory:
+    """
+    Factory to build the AWS client.
+
+    Boto client creation is resource intensive. This class caches all Boto
+    clients it creates and must be used instead of directly using boto lib.
+    """
+
+    def __init__(
+        self,
+        use_ssl: bool = False,
+        verify: bool = False,
+    ):
+        """
+        :param use_ssl: Whether to use SSL
+        :param verify: Whether to verify SSL certificates
+        """
+        self._use_ssl = use_ssl
+        self._verify = verify
+        self._session = Session()
+        self._config = Config(max_pool_connections=MAX_POOL_CONNECTIONS)
+
+    def __call__(self, *args, **kwargs) -> BaseClient:
+        """
+        Every `ConnectFactory` instance is a proxy for creating the internal client.
+        """
+        return self.get_internal_client(*args, **kwargs)
+
+    def get_internal_client(
+        self,
+        service_name: str,
+        region_name: str,
+        endpoint_url: str = None,
+        aws_access_key_id: str = INTERNAL_AWS_ACCESS_KEY_ID,
+        aws_secret_access_key: str = INTERNAL_AWS_SECRET_ACCESS_KEY,
+        config: Config = None,
+    ) -> BaseClient:
+        """
+        Build and return client for connections originating within LocalStack.
+
+        Region in `_TargetArn` takes precedence when specified.
```

Comment:
This will happen serverside. The `RegionContextEnricher` will set the region from the DTO:TargetArn if available, then resort to standard AWS headers.

This approach is due to current design limitations of how we're leveraging Boto's event system